core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -0,0 +1,180 @@
+import java.util.List;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.transforms.PartitionSpecVisitor;
+
+public class Partitioning {
+  private Partitioning() {
+  }
+
+  /**
+   * Check whether the spec contains a bucketed partition field.
+   *
+   * @param spec a partition spec
+   * @return true if the spec has field with a bucket transform
+   */
+  public static boolean hasBucketField(PartitionSpec spec) {
+    List<Boolean> bucketList = PartitionSpecVisitor.visit(spec, new PartitionSpecVisitor<Boolean>() {
+      @Override
+      public Boolean identity(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean bucket(int fieldId, String sourceName, int sourceId, int width) {
+        return true;
+      }
+
+      @Override
+      public Boolean truncate(int fieldId, String sourceName, int sourceId, int width) {
+        return false;
+      }
+
+      @Override
+      public Boolean year(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean month(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean day(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean hour(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean alwaysNull(int fieldId, String sourceName, int sourceId) {
+        return false;
+      }
+
+      @Override
+      public Boolean unknown(int fieldId, String sourceName, int sourceId, String transform) {
+        return false;
+      }
+    });
+
+    return bucketList.stream().anyMatch(Boolean::booleanValue);
+  }
+
+  /**
+   * Create a sort order that will group data for a partition spec.
+   * <p>
+   * If the partition spec contains bucket columns, the sort order will also have a field to sort by a column that is
+   * bucketed in the spec. The column is selected by the highest number of buckets in the transform.
+   *
+   * @param spec a partition spec
+   * @return a sort order that will cluster data for the spec
+   */
+  public static SortOrder sortOrderFor(PartitionSpec spec) {
+    if (spec.isUnpartitioned()) {
+      return SortOrder.unsorted();
+    }
+
+    SortOrder.Builder builder = SortOrder.builderFor(spec.schema());
+    SpecToOrderVisitor converter = new SpecToOrderVisitor(builder);
+    PartitionSpecVisitor.visit(spec, converter);
+
+    // columns used for bucketing are high cardinality; add one to the sort at the end
+    String bucketColumn = converter.bucketColumn();
+    if (bucketColumn != null) {
+      builder.asc(bucketColumn);
+    }
+
+    return builder.build();
+  }
+
+  private static class SpecToOrderVisitor implements PartitionSpecVisitor<Void> {
+    private final SortOrder.Builder builder;
+    private String bucketColumn = null;
```

Comment:
Are we guaranteed to have only one bucket column?  I actually didn't think of this until now, but Hive allows for bucketing on multiple columns (though those are combined and hashed I believe.  However, I didn't immedaitely see anything that prevents someone from defining multiple bucket columns.

Reply:
There's no reason why we can't have more than one bucket transform, but Iceberg doesn't mix fields together in bucket transforms. The reason is that each bucket column should be useful independently.

For example, you might have `PARTITION BY (bucket(8, user_id), bucket(16, item_id))`. Joining to both users and items independently could take advantage of those bucketed columns. And joining two tables with `user_id` and `item_id` (like `purchases` and `reviews`) could use a combination of both: `user_bucket << 16 | item_bucket`.

Reply:
I guess I wasn't clear on my point.  I understand why we're not doing what Hive does in this case, but rather how we are only using a single bucket in the sort order above on line 106.

Reply:
Oh, sorry about that. The bucket column is tracked because it is assumed to be a high cardinality column. Since we want to end a sort order with a high cardinality column, we keep track of the bucket column with the largest number of buckets and add that to the inferred sort order.

```java
      // the column with highest cardinality is usually the one with the highest number of buckets
      if (numBuckets > highestNumBuckets) {
        this.highestNumBuckets = numBuckets;
        this.bucketColumn = sourceName;
      }
```

Then above in `sortOrderFor(PartitionSpec)`, if the bucket column is present, it is added to the order.

Reply:
Ah, ok.  That makes sense.